core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala

```diff
@@ -1399,6 +1397,158 @@ class AdminClientIntegrationTest extends IntegrationTestHarness with Logging {
+    assertEquals((Defaults.LogRetentionHours * 60 * 60 * 1000).toString, configs.get(topicResource1).get(LogConfig.RetentionMsProp).value)
+
+    assertEquals("0.9", configs.get(topicResource2).get(LogConfig.MinCleanableDirtyRatioProp).value)
+    assertEquals("lz4", configs.get(topicResource2).get(LogConfig.CompressionTypeProp).value)
+
+    //verify config value substract
+    topicConfigEntries1 = Seq(
+      new AlterConfigOp(new ConfigEntry(LogConfig.CleanupPolicyProp, LogConfig.Compact), AlterConfigOp.OpType.SUBSTRACT)
+    ).asJava
+
+   alterResult = client.incrementalAlterConfigs(Map(
+      topicResource1 -> topicConfigEntries1
+    ).asJava)
+    alterResult.all.get
+
+    // Verify that topics were updated correctly
+    describeResult = client.describeConfigs(Seq(topicResource1).asJava)
+    configs = describeResult.all.get
+
+    assertEquals("delete", configs.get(topicResource1).get(LogConfig.CleanupPolicyProp).value)
+
+    // Alter topics with validateOnly=true
+    topicConfigEntries1 = Seq(
+      new AlterConfigOp(new ConfigEntry(LogConfig.CleanupPolicyProp, LogConfig.Compact), AlterConfigOp.OpType.APPEND)
+    ).asJava
+
+    alterResult = client.incrementalAlterConfigs(Map(
+      topicResource1 -> topicConfigEntries1
+    ).asJava, new IncrementalAlterConfigsOptions().validateOnly(true))
+    alterResult.all.get
+
+    // Verify that topics were not updated due to validateOnly = true
+    describeResult = client.describeConfigs(Seq(topicResource1).asJava)
+    configs = describeResult.all.get
+
+    assertEquals("delete", configs.get(topicResource1).get(LogConfig.CleanupPolicyProp).value)
+
+    //Alter topics with validateOnly=true with invalid configs
+    topicConfigEntries1 = Seq(
+      new AlterConfigOp(new ConfigEntry(LogConfig.CompressionTypeProp, "zip"), AlterConfigOp.OpType.SET)
+    ).asJava
+
+    alterResult = client.incrementalAlterConfigs(Map(
+      topicResource1 -> topicConfigEntries1
+    ).asJava, new IncrementalAlterConfigsOptions().validateOnly(true))
+
+    assertFutureExceptionTypeEquals(alterResult.values().get(topicResource1), classOf[InvalidRequestException],
+      Some("Invalid config value for resource"))
+  }
+
+  @Test
+  def testInValidIncrementalAlterConfigs(): Unit = {
+    client = AdminClient.create(createConfig)
+
+    // Create topics
+    val topic1 = "incremental-alter-configs-topic-1"
+    val topicResource1 = new ConfigResource(ConfigResource.Type.TOPIC, topic1)
+    createTopic(topic1)
+
+    val topic2 = "incremental-alter-configs-topic-2"
+    val topicResource2 = new ConfigResource(ConfigResource.Type.TOPIC, topic2)
+    createTopic(topic2)
+
+    var topicConfigEntries1 = Seq(
+      new AlterConfigOp(new ConfigEntry(LogConfig.CleanupPolicyProp, LogConfig.Delete), AlterConfigOp.OpType.APPEND),
+      new AlterConfigOp(new ConfigEntry(LogConfig.CleanupPolicyProp, LogConfig.Compact), AlterConfigOp.OpType.DELETE)
```

Comment:
I think that we should also add a new, **valid** config entry here and ensure that it gets updated despite the invalid `cleanup.policy`. That's how I understand the KIP should work at least.

If that's not the intention, maybe we could add a valid config entry and assert that it doesn't get updated.

Reply:
We don't do partial update for a resource. either we apply all configs or fail the request. 
Updated the test to add a valid config entry and assert that it doesn't get updated.

Reply:
The intention in KIP-339 is to allow partial updates. Each AlterConfigOp is independent and just because one fails, doesn't mean any other one will fail (except in the case of duplicate keys as described earlier).

Reply:
@cmccabe Updated the PR. There are scenarios, where we need to update/validate related configs in single batch. In this cases, each AlterConfigOp are not  independent .
Example: while adding listener, user want may want to update listener and listener.security.protocol.map configs Without listener.security.protocol.map config, listener config update may fail. 

I have added a check to decide if we need batch config validation or not. Not sure if this is right approach. Please take a look when you get a chance.

cc @rajinisivaram